pypuf/learner/evolution_strategies/becker.py

```diff
@@ -0,0 +1,311 @@
+            self.step_size = self.update_ss(
+                self.step_size, self.c_sigma, self.d_sigma, self.path_ss)
+        return self.new_LTF
+
+
+    # updating methods of evolution strategies
+    @staticmethod
+    def reproduce(mean, cov_matrix, pop_size, step_size):
+        # returns a new generation of individuals as 2D array (pop_size, n)
+        mutations = np.random.multivariate_normal(np.zeros(np.shape(mean)),
+                                                  cov_matrix, pop_size)
+        duplicated_mean = np.tile(np.matrix(mean), (pop_size, 1))
+        return duplicated_mean + (step_size * mutations)
+
+    @staticmethod
+    def update_mean(mean, step_size, parent):
+        # returns mean of a new population as array (n)
+        return mean + step_size*parent
+
+    @staticmethod
+    def cumulation_for_cm(path_cm, c_c, path_ss, n, mu_w, parent):
+        # returns cumulated evolution path of covariance matrix
+        path_cm = path_cm * (1-c_c)
+        if(np.linalg.norm(path_ss) < 1.5 * np.sqrt(n)):
+            path_cm = path_cm + (np.sqrt(1 - (1-c_c)**2) *
+                                 np.sqrt(mu_w) * parent)
+        return path_cm
+
+    @staticmethod
+    def cumulation_for_ss(path_ss, c_sigma, mu_w, cov_matrix, parent):
+        # returns cumulated evolution path of step-size
+        cm_eigen_dec = Reliability_based_CMA_ES.modify_eigen_decomposition(cov_matrix)
+        return (1-c_sigma) * path_ss + np.sqrt(1 - (1-c_sigma)**2) * \
+                                       np.sqrt(mu_w) * cm_eigen_dec @ parent
+
+    @staticmethod
+    def update_cm(cov_matrix, c_1, c_mu, path_cm, cm_mu):
+        # returns covariance matrix of a new population (pop_size, pop_size)
+        return (1 - c_1 - c_mu) * cov_matrix + c_1 * path_cm * path_cm.T \
+               + c_mu * cm_mu
+
+    @staticmethod
+    def update_ss(step_size, c_sigma, d_sigma, path_ss):
+        # returns step-size of a new population
+        factor = np.exp((c_sigma / d_sigma) *
+                        ((np.linalg.norm(path_ss) / np.sqrt(np.shape(path_ss)[1])) - 1))
+        return step_size * factor
+
+    def fitness(self, challenges, challenge_num, measured_rels, individuals, epsilon, epsilon_theo):
+        # returns individuals sorted by their fitness
+        pop_size = np.shape(individuals)[0]
+        built_LTFs = Reliability_based_CMA_ES.build_LTFArrays(individuals)
+        delay_diffs = Reliability_based_CMA_ES.get_delay_differences(
+            built_LTFs, pop_size, challenges, challenge_num)
+        reliabilities = Reliability_based_CMA_ES.get_reliabilities(delay_diffs, epsilon_theo)
+        correlations = Reliability_based_CMA_ES.get_correlations(reliabilities, measured_rels)
+        for i in range(np.shape(correlations)):
+            if correlations[i] > 0.9*epsilon:
+                self.new_LTF = individuals[i, :]
+                self.terminate = True
```

Comment:
This will cause a `break` in the main while loop of the learning algorithm, which is kind of unusual. I think it is easier to read if this for loop here is moved to the main while loop entirely, then no class variable `self.terminate` is needed; while maintaining the same run time and functionality.